src/Equinox.MessageDb/MessageDbClient.fs

```diff
@@ -0,0 +1,126 @@
+            let events = ResizeArray()
+            let mutable hasNext = hasNext
+
+            while hasNext do
+                let meta =
+                    if reader.IsDBNull(3) then
+                        "null"
+                    else
+                        reader.GetString(3)
+
+                let time =
+                    DateTime.SpecifyKind(reader.GetDateTime(7), DateTimeKind.Utc)
+
+                let timestamp = DateTimeOffset(time)
+
+                let timelineEvent =
+                    TimelineEvent.Create(
+                        index = reader.GetInt64(0),
+                        eventType = reader.GetString(1),
+                        data = JsonSerializer.Deserialize<JsonElement>(reader.GetString(2)),
+                        meta = JsonSerializer.Deserialize<JsonElement>(meta),
+                        eventId = reader.GetGuid(4),
+                        ?correlationId =
+                            (if reader.IsDBNull(5) then
+                                 None
+                             else
+                                 Some(reader.GetString(5))),
+                        ?causationId =
+                            (if reader.IsDBNull(6) then
+                                 None
+                             else
+                                 Some(reader.GetString(6))),
+                        timestamp = timestamp
+                    )
+
+                events.Add(timelineEvent)
+                let! next = reader.ReadAsync(ct)
+                hasNext <- next
+
+            return events.ToArray()
+        }
+
+    member private _.PrepareWriteCommand
+        (cmd: NpgsqlBatchCommand)
+        (streamName: string)
+        version
+        (message: IEventData<JsonElement>)
+        =
+        cmd.CommandText <- "select 1 from write_message(@Id::text, @StreamName, @EventType, @Data, @Meta, @ExpectedVersion)"
+
+        cmd.Parameters.AddWithValue("Id", NpgsqlDbType.Uuid, message.EventId) |> ignore
+        cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, streamName) |> ignore
+        cmd.Parameters.AddWithValue("EventType", NpgsqlDbType.Text, message.EventType) |> ignore
+        let data, meta = message.Data, message.Meta
+        cmd.Parameters.AddWithValue("Data", NpgsqlDbType.Jsonb, data.GetRawText()) |> ignore
+        let meta =
+            match meta.ValueKind with
+            | JsonValueKind.Null | JsonValueKind.Undefined -> "null"
+            | _ -> meta.GetRawText()
+        cmd.Parameters.AddWithValue("Meta", NpgsqlDbType.Jsonb, meta) |> ignore
+        cmd.Parameters.AddWithValue("ExpectedVersion", NpgsqlDbType.Bigint, version) |> ignore
+
+        cmd
+
+    member __.WriteMessages(streamName, events, version: int64, ct) =
+        task {
+            try
+                use! conn = source ()
+                use transaction = conn.BeginTransaction()
+                use batch = new NpgsqlBatch(conn, transaction)
```

Comment:
~I guess its v premature optimisation but given writing one event is the common case, maybe special case that without a batch~ What am I saying, DONT DO THAT!?